File: frontend/src/types/stats.ts
export type DailyPoint = {
  day: string;
  sessions?: number;
  avg_bandwidth_bps?: number;
};

export type SourceCount = {
  source: "tautulli" | "sftpgo" | string;
  sessions: number;
};

export type OverviewStats = {
  total_sessions: number;
  active_sessions: number;
  ended_sessions: number;
  stale_sessions: number;
  total_shared_bytes: number;
  total_shared_human: string;
  sessions_by_day: Array<{ day: string; sessions: number }>;
  sessions_by_week: Array<{ day: string; sessions: number }>;
  sessions_by_month: Array<{ day: string; sessions: number }>;
  sessions_by_year: Array<{ day: string; sessions: number }>;
  bandwidth_by_day: Array<{ day: string; avg_bandwidth_bps: number }>;
  bandwidth_by_week: Array<{ day: string; avg_bandwidth_bps: number }>;
  bandwidth_by_month: Array<{ day: string; avg_bandwidth_bps: number }>;
  bandwidth_by_year: Array<{ day: string; avg_bandwidth_bps: number }>;
  source_distribution: SourceCount[];
  active_by_source: SourceCount[];
  play_count_by_weekday: Array<{ label: string; sessions: number }>;
  play_count_by_hour: Array<{ hour: number; sessions: number }>;
  play_count_by_platform: Array<{ label: string; sessions: number }>;
  play_count_by_media_type: Array<{ label: string; sessions: number }>;
};

export type TopUser = {
  user_name: string;
  sessions: number;
  active_sessions: number;
  avg_bandwidth_bps: number | null;
  last_seen_at: string | null;
};

export type UsersStatsResponse = {
  items: TopUser[];
};

export type TopMediaItem = {
  title: string;
  media_type: string;
  sessions: number;
  unique_users: number;
  avg_bandwidth_bps: number | null;
  sample_session_id: number | null;
};

export type MediaStatsResponse = {
  top_movies: TopMediaItem[];
  top_series: TopMediaItem[];
};

export type UserInsightItem = {
  user_name: string;
  total_sessions: number;
  movie_sessions: number;
  episode_sessions: number;
  total_watch_hours: number;
  movie_watch_hours: number;
  episode_watch_hours: number;
  unique_titles_monthly: number;
  unique_movies_monthly: number;
  unique_series_monthly: number;
  last_seen_at: string | null;
};

export type UserInsightsResponse = {
  items: UserInsightItem[];
  leaders: {
    most_sessions_user: { user_name: string; value: number };
    most_watch_hours_user: { user_name: string; value: number };
    most_movies_user: { user_name: string; value: number };
    most_series_user: { user_name: string; value: number };
  };
  peak_hours: Array<{ hour: number; sessions: number }>;
  timezone: string;
  play_count_rule: string;
};

File: backend/app/api/v1/schemas/stats.py
from datetime import datetime

from pydantic import BaseModel


class SourceCount(BaseModel):
    source: str
    sessions: int


class DailySessionsPoint(BaseModel):
    day: str
    sessions: int


class DailyBandwidthPoint(BaseModel):
    day: str
    avg_bandwidth_bps: float


class LabeledSessionsPoint(BaseModel):
    label: str
    sessions: int


class HourSessionsPoint(BaseModel):
    hour: int
    sessions: int


class DailySharedPoint(BaseModel):
    day: str
    shared_bytes: int


class HourSharedPoint(BaseModel):
    hour: int
    shared_bytes: int


class OverviewStatsResponse(BaseModel):
    total_sessions: int
    active_sessions: int
    ended_sessions: int
    stale_sessions: int
    total_shared_bytes: int
    total_shared_human: str
    sessions_by_day: list[DailySessionsPoint]
    sessions_by_week: list[DailySessionsPoint]
    sessions_by_month: list[DailySessionsPoint]
    sessions_by_year: list[DailySessionsPoint]
    bandwidth_by_day: list[DailyBandwidthPoint]
    bandwidth_by_week: list[DailyBandwidthPoint]
    bandwidth_by_month: list[DailyBandwidthPoint]
    bandwidth_by_year: list[DailyBandwidthPoint]
    source_distribution: list[SourceCount]
    active_by_source: list[SourceCount]
    play_count_by_weekday: list[LabeledSessionsPoint]
    play_count_by_hour: list[HourSessionsPoint]
    play_count_by_platform: list[LabeledSessionsPoint]
    play_count_by_media_type: list[LabeledSessionsPoint]
    shared_by_day: list[DailySharedPoint]
    shared_by_week: list[DailySharedPoint]
    shared_by_month: list[DailySharedPoint]
    shared_by_year: list[DailySharedPoint]
    shared_by_hour: list[HourSharedPoint]


class TopUserStat(BaseModel):
    user_name: str
    sessions: int
    active_sessions: int
    avg_bandwidth_bps: float | None
    last_seen_at: datetime | None


class TopUsersResponse(BaseModel):
    items: list[TopUserStat]


class TopMediaItem(BaseModel):
    title: str
    media_type: str
    sessions: int
    unique_users: int
    avg_bandwidth_bps: float | None
    sample_session_id: int | None


class TopMediaResponse(BaseModel):
    top_movies: list[TopMediaItem]
    top_series: list[TopMediaItem]


class UserInsightsItem(BaseModel):
    user_name: str
    total_sessions: int
    movie_sessions: int
    episode_sessions: int
    total_watch_hours: float
    movie_watch_hours: float
    episode_watch_hours: float
    unique_titles_monthly: int
    unique_movies_monthly: int
    unique_series_monthly: int
    last_seen_at: datetime | None


class UserInsightsLeader(BaseModel):
    user_name: str
    value: float


class UserInsightsLeaders(BaseModel):
    most_sessions_user: UserInsightsLeader
    most_watch_hours_user: UserInsightsLeader
    most_movies_user: UserInsightsLeader
    most_series_user: UserInsightsLeader


class PeakHourPoint(BaseModel):
    hour: int
    sessions: int


class UserInsightsResponse(BaseModel):
    items: list[UserInsightsItem]
    leaders: UserInsightsLeaders
    peak_hours: list[PeakHourPoint]
    timezone: str
    play_count_rule: str
